llvm/test/Transforms/SROA/select-gep.ll

```diff
@@ -202,8 +206,10 @@ define i32 @test_select_idx_escaped(i1 %c, ptr %p) {
 ; CHECK-NEXT:    [[GEP1:%.*]] = getelementptr inbounds i8, ptr [[ALLOCA]], i64 24
 ; CHECK-NEXT:    store i32 2, ptr [[GEP1]], align 4
 ; CHECK-NEXT:    [[IDX:%.*]] = select i1 [[C:%.*]], i64 24, i64 0
-; CHECK-NEXT:    [[GEP2:%.*]] = getelementptr inbounds i8, ptr [[ALLOCA]], i64 [[IDX]]
-; CHECK-NEXT:    [[RES:%.*]] = load i32, ptr [[GEP2]], align 4
+; CHECK-NEXT:    [[DOTSROA_GEP:%.*]] = getelementptr inbounds i8, ptr [[ALLOCA]], i64 24
+; CHECK-NEXT:    [[DOTSROA_GEP1:%.*]] = getelementptr inbounds i8, ptr [[ALLOCA]], i64 0
+; CHECK-NEXT:    [[IDX_SROA_SEL:%.*]] = select i1 [[C]], ptr [[DOTSROA_GEP]], ptr [[DOTSROA_GEP1]]
+; CHECK-NEXT:    [[RES:%.*]] = load i32, ptr [[IDX_SROA_SEL]], align 4
```

Comment:
Ideally, we would not transform this case (where no splitting is possible), but that's a general/pre-existing issue with how SROA does select unfolding.

Reply:
Yup. In practice, we have instcombine+SROA in multiple places in the pipeline, so next instcombine will probably merge these GEPs again.

Reply:
Looks like 'transactional' way of transforming basic blocks or functions might be a useful concept. 
E.g. a pass may attempt to do some transformation but does not know ahead of time if it will succeed. If it works out, keep the new block, if not, restore the original version.
I'm surprised we do not have it as a standard feature in LLVM. Or, maybe we do?  In any case, that's a subject for a separate conversation.